connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1181,12 +1186,17 @@ private boolean startConnector(String connectorName) {
         final ConnectorContext ctx = new HerderConnectorContext(this, connectorName);
         final TargetState initialState = configState.targetState(connectorName);
         boolean started = worker.startConnector(connectorName, configProps, ctx, this, initialState);
+        final Map<String, String> clientAndConverterConfigs = clientAndConverterConfigs(configProps);
+        final Map<String, String> priorClientAndConverterConfigs =
+            connectorClientAndConverterConfigs.put(connectorName, clientAndConverterConfigs);
 
         // Immediately request configuration since this could be a brand new connector. However, also only update those
         // task configs if they are actually different from the existing ones to avoid unnecessary updates when this is
         // just restoring an existing connector.
-        if (started && initialState == TargetState.STARTED)
-            reconfigureConnectorTasksWithRetry(time.milliseconds(), connectorName);
+        if (started && initialState == TargetState.STARTED) {
+            boolean forceReconfigure = !clientAndConverterConfigs.equals(priorClientAndConverterConfigs);
+            reconfigureConnectorTasksWithRetry(time.milliseconds(), connectorName, forceReconfigure);
```

Comment:
Unlike in the standalone, it's a bit more difficult to check how the connector configuration has changed. This `startConnector(...)` method seems to be called in two places: from `restartConnector(...)` and `processConnectorConfigUpdates`, which IIRC happens whenever the connector configuration has changed in some way. 

I'm wondering how often a connector might not change it's task configs when the connector config has changed. What do you think about just always restarting the tasks even if the generated task configs have not changed?

WDYT?

@kkonstantine, the `processConnectorConfigUpdates(...)` method is called during rebalances, and maybe I'm missing cases where the recent rebalance improvements handle the task configs more frequently than I recall.

Reply:
This method is invoked any time a connector has to be started, which in addition to the cases you've noted, also includes starting newly-assigned connectors after a rebalance.

Writing the task configs to the config topic would break any cluster using eager rebalancing, since task config updates cause a rebalance. The cluster would enter an endless loop where the connector is started, its task configs are requested and then written to the config topic, the new task configs are picked up and trigger a rebalance, which causes the connector to be stopped, then started, and the whole thing begins again.

This is a low-impact bug and, on its own, probably not worth the effort it would take to be able to write unnecessary task configs to the config topic without breaking clusters using the eager rebalancing protocol.